### DISCUSSION TITLE
How to check you haven't used an undefined variable in the template?

### DISCUSSION THREAD

**Discussion by coretl** [copier-org/copier#1512] [Q&A]
I've been bitten a couple of times by this workflow:
- Change the name of a variable in `copier.yml`
- Forget to rename some of the references to that variable in my template
- End up with that variable being rendered blank in a text file somewhere when using the template

It would appear that Jinja does this by default, and the way it suggests to raise an error is by supplying [`undefined=StrictUndefined`](https://jinja.palletsprojects.com/en/3.1.x/api/#jinja2.StrictUndefined) in its [`Environment`](https://jinja.palletsprojects.com/en/3.1.x/api/#jinja2.Environment). I tried this in `copier.yml`:
```yaml
_envops:
    undefined: StrictUndefined
```

But it fails because it can't turn the string into a callable:
```python
  File "/venv/lib/python3.11/site-packages/copier/main.py", line 1024, in run_copy
    with Worker(src_path=src_path, dst_path=Path(dst_path), **kwargs) as worker:
  File "/venv/lib/python3.11/site-packages/copier/main.py", line 205, in __exit__
    raise value
  File "/venv/lib/python3.11/site-packages/copier/main.py", line 1025, in run_copy
    worker.run_copy()
  File "/venv/lib/python3.11/site-packages/copier/main.py", line 745, in run_copy
    self._ask()
  File "/venv/lib/python3.11/site-packages/copier/main.py", line 421, in _ask
    jinja_env=self.jinja_env,
              ^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/functools.py", line 1001, in __get__
    val = self.func(instance)
          ^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/copier/main.py", line 500, in jinja_env
    env = SandboxedEnvironment(
          ^^^^^^^^^^^^^^^^^^^^^
  File "/venv/lib/python3.11/site-packages/jinja2/sandbox.py", line 253, in __init__
    super().__init__(*args, **kwargs)
  File "/venv/lib/python3.11/site-packages/jinja2/environment.py", line 366, in __init__
    _environment_config_check(self)
  File "/venv/lib/python3.11/site-packages/jinja2/environment.py", line 126, in _environment_config_check
    assert issubclass(
           ^^^^^^^^^^^
TypeError: issubclass() arg 1 must be a class
```

Is there a way to make the `copier copy` command fail if you use an undefined name in a template?

Related to #313 but with a different aim in mind

**Comment by yajo** [comment]
Copier is lacking this feature, but it seems like a good idea. I opened https://github.com/copier-org/copier/issues/1522 to track it.

**Comment by object-Object** [comment]
It would be great for this to be supported by Copier, but here are a couple of workarounds in the meantime.

## 1: Monkeypatch `Worker`

This method only applies `StrictUndefined` when the patched script is used (ie. during development), not when normally copying from the template; this means that end users won't have to install any extra packages or change how they use the template.

Create a file (eg. `copier_strict.py`) with these contents, and run that script instead of `copier` (eg. `python -m copier_strict copy path/to/project/template path/to/destination`). For programmatic use, `StrictUndefinedWorker.run_copy` could be used directly (basically reimplementing `copier.run_copy`) instead of patching Copier.

```py
from functools import cached_property
import copier.main
from jinja2 import StrictUndefined

class StrictUndefinedWorker(copier.main.Worker):
    @cached_property
    def jinja_env(self):
        env = super().jinja_env
        env.undefined = StrictUndefined
        return env

copier.main.Worker = StrictUndefinedWorker

if __name__ == "__main__":
    from copier.cli import CopierApp

    CopierApp.run()
```

## 2: Custom extension

This method requires you to create a Python package that end users can install, since it uses the `_jinja_extensions` option.

Create a package with these contents, install it, and add `your_package.StrictUndefinedExtension` to `_jinja_extensions` in your `copier.yml` file. Tell your end users to install this package before copying the template.

```py
from jinja2 import Environment, StrictUndefined
from jinja2.ext import Extension

class StrictUndefinedExtension(Extension):
    def __init__(self, environment: Environment):
        super().__init__(environment)
        environment.undefined = StrictUndefined
```
